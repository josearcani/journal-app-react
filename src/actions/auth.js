import { getAuth, signInWithPopup } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

// async action
export const startLoginEmailPassword = (email, password) => {
  // use fetch or axios to get data

  return (dispatch) => {
    setTimeout(() => {
      dispatch( login('asfasi23534on32', 'Gerardo') )
    }, 1500);
  }
}

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
      })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})
