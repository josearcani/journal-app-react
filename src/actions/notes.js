import { addDoc, collection } from "firebase/firestore";
import { db } from "../firebase/firebase-config";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    // { auth, notes, ui }
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    }
    
    const docRef = await addDoc(collection(db ,`${uid}/journal/notes`), newNote);

    console.log(docRef);
  }
}