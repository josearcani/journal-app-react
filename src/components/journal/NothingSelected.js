import React from 'react'

export const NothingSelected = () => {
  return (
    <div className="nothing__main-context">
      <p>New entry</p>

      <i className="fas fa-plus fa-4x mt-5"></i>
      
    </div>
  )
}
