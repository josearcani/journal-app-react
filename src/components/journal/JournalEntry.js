import React from 'react';
import { useDispatch } from 'react-redux';
import moment from 'moment';
import { noteActive } from '../../actions/notes';

export const JournalEntry = ({ id, title, body, url, date }) => {

  const noteDate = moment(date);
  const dispatch = useDispatch();

  const handleEntryClick = () => {
    dispatch(noteActive(id, { title, body, date, url }));
  }

  return (
    <div
      className="journal__entry animate__animated animate__fadeInLeft animate__faster"
      onClick={ handleEntryClick }
    >

      {
        url && <div 
        className="journal__entry-picture"
        style={{
          backgroundSize: 'cover',
          backgroundImage: `url(${ url })`,
          backgroundRepeat: 'no-repeat',
          backgroundPosition: 'center'
        }}
        ></div>
      }

      <div className="journal__entry-body">
        <p className="journal__entry-title">
          { title }
        </p>
        <p className="journal__entry-content">
          { body }
        </p>

      </div>

      <div className="journal__entry-date-box">
        <span>{ noteDate.format('dddd')}</span>
        <h4>{ noteDate.format('Do') }</h4>
      </div>


    </div>
  )
}
