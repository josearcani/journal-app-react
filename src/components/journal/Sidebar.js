import React from 'react';
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';

import { JournalEntries } from './JournalEntries';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

export const Sidebar = () => {

  const dispatch = useDispatch();

  // useSelector gets the state from store
  const { name } = useSelector(state => state.auth)

  const handleLogout = () => {
    dispatch(startLogout())
    // later clean notes and journal for current user
  }

  const handleAddNew = () => {
    dispatch(startNewNote());
  }

  return (
    <aside className="journal__sidebar">

      <div className="journal__sidebar-navbar">
        <h3 className="mt-1">
          <i className="fas fa-book" />
          <span>{ name }</span>
        </h3>

        {/* dropdown to implement */}
        <button
          className="btn btn-danger"
          onClick={ handleLogout }
        >
          <i className="fas fa-sign-out-alt"></i>&nbsp;
          Logout
        </button>

      </div>

      <div className="journal__forecast">
        <div className="journal__forecast-datetime">
          <p>Friday</p>
          <span >January 13, 2022</span>
        </div>
        <div className="journal__forecast-temperature">
          <p className="unit">
            <i className="far fa-sun"></i>&nbsp;34<i>&deg;</i>
          </p>
          <span>Seul</span>
        </div>

      </div>

      <div
        className="journal__new-entry"
        onClick={ handleAddNew }
      >
        <i className="fas fa-sticky-note" />&nbsp;
        <p>New entry</p>
      </div>


      <JournalEntries />

    </aside>
  )
}
