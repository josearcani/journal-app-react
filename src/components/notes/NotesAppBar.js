import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

  const dispatch = useDispatch();

  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote(active));
  }

  const handleUploadPicture = () => {
    document.querySelector('#fileSelector').click();
  }

  const handleFileChange = (e) => {
    // e.target => <input />
    // e.target.files ==> FileList {0: File, length: 1}
    const file = e.target.files[0];
    dispatch(startUploading(file));
  }

  return (
    <div className="notes__appbar">
      <span>28 de septiembre 2021</span>

      <input
        id="fileSelector"
        type="file"
        name="file"
        style={{ display: "none" }}
        onChange={ handleFileChange }
      />

      <div>
        <button
          className="btn"
          onClick={ handleUploadPicture }
        >
          Picture
        </button>

        <button
          className="btn"
          onClick={ handleSave }
        >
          Save
        </button>
      </div>
    </div>
  )
}
