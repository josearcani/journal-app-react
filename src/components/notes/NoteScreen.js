import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import { noteActive, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {

  const dispatch = useDispatch();
  const { active: note } = useSelector(state => state.notes);
  const [ formValues, handleInputChange, cleanForm ] = useForm(note);
  
  const activeId = useRef(note.id);
  // this ref is just for the first note

  useEffect(() => {
    if (activeId.current !== note.id) {
      cleanForm(note);
      activeId.current = note.id;
      // update the new id to reference and clean with new active note
    }
  }, [ note, cleanForm ])


  useEffect(() => {
    dispatch(noteActive(formValues.id, { ...formValues }))
  }, [ dispatch ,formValues ])

  const handleDelete = () => {
    dispatch(startDeleting(formValues.id))
  }

  return (
    <div className="notes__main-content">

      <NotesAppBar />

      <div className="notes__content">

        <input
          type="text"
          placeholder="Some delightful thought"
          className="notes__title-input"
          autoComplete="off"
          name="title"
          onChange={ handleInputChange }
          value={ formValues.title }
        />

        <textarea
          placeholder="Hey write over here..."
          className="notes__textarea"
          name="body"
          onChange={ handleInputChange }
          value={ formValues.body }
        ></textarea>


        {
          note.url &&
              (<div className="notes__image">
                <img
                  src={ note.url }
                  alt={ formValues.title }
                />
              </div>)
        }

      </div>

        <button
          className="btn btn-danger"
          onClick={ handleDelete }
        >
          Delete
        </button>

    </div>
  )
}
