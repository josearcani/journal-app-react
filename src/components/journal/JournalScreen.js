import React from 'react';
import { useSelector } from 'react-redux';
import { NoteScreen } from '../notes/NoteScreen';
import { Sidebar } from './Sidebar';
import { NothingSelected } from './NothingSelected';

export const JournalScreen = () => {

  const { active } = useSelector(state => state.notes);
  const { toggle } = useSelector(state => state.ui);
  
  // const handleToogle = () => {
  //   console.log('asdfasdf');
  // }
  console.log(toggle);

  return (
    <div className="journal__main-context animate__animated animate__fadeIn animate__faster">

      <Sidebar />

      <main className={`journal__main ${toggle && 'is-full-width'}`}>

        {
          active
            ? <NoteScreen />
            : <NothingSelected />
        }

      </main>
    </div>
  )
}
