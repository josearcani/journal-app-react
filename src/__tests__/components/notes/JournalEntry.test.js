import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { JournalEntry } from '../../../components/journal/JournalEntry';
import { noteActive } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const notes = {
  id: 123,
  title: 'Test title',
  body: 'I am a body',
  url: 'https://somehow.test/noPicture.jpg',
  date: 8998989
}

const wrapper = mount(
  <Provider store={ store }>
    <JournalEntry { ...notes } />
  </Provider>
);

describe('Test on <JournalEntry />', () => {
  test('should JournalEntry render normally', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should call activeNote action', () => {
    wrapper.find('.journal__entry').simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(
      noteActive(notes.id, { ...notes })
    )
  });
});
