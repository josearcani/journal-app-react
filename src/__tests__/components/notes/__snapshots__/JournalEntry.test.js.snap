// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <JournalEntry /> should JournalEntry render normally 1`] = `
<div
  className="journal__entry animate__animated animate__fadeInLeft animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundColor": "cover",
        "backgroundImage": "url(https://somehow.test/noPicture.jpg)",
        "backgroundPosition": "center",
        "backgroundRepeat": "no-repeat",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Test title
    </p>
    <p
      className="journal__entry-content"
    >
      I am a body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
